Algorithms/Search/FastSearcher.cs

```diff
@@ -0,0 +1,105 @@
+﻿using System;
+
+namespace Algorithms.Search
+{
+    /// <summary>
+    /// The idea: you could combine the advantages from both binary-search and interpolation search algorithm.
+    /// Time complexity:
+    ///     worst case: Item couldn't be found: O(log n),
+    ///     average case: O(log log n),
+    ///     best case: O(1).
+    /// Note: This algorithm is recursive and the array has to be sorted beforehand.
+    /// </summary>
+    public class FastSearcher
+    {
+        private int[] array;
+
+        /// <summary>
+        /// Finds the index of the given item in the array.
+        /// </summary>
+        /// <param name="array">Defines the span of numbers.</param>
+        /// <param name="item">The number to be found.</param>
+        /// <returns>Returns the index of the item in the array.</returns>
+        public int FindIndex(int[] array, int item)
+        {
+            this.array = array;
+            int from = 0, to = array.Length - 1;
+            return FindIndex(from, to, item);
+        }
+
+        /// <summary>
+        /// Finds index of first item in array that satisfies specified term
+        /// throws ItemNotFoundException if the item couldn't be found.
+        /// </summary>
+        /// <param name="sectionStartIndex">Defines the lower boundary of the section.</param>
+        /// <param name="sectionEndIndex">Defines the upper boundary of the section.</param>
+        /// <param name="item">Term to check against.</param>
+        /// <returns>Index of first item that satisfies term.</returns>
+        private int FindIndex(int sectionStartIndex, int sectionEndIndex, int item)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Refactor this method to reduce its Cognitive Complexity from 17 to the 15 allowed.](https://app.codacy.com/app/siriak/C-Sharp/pullRequest?prid=4578957)